iohub/dataset_base.py

```diff
@@ -0,0 +1,88 @@
+from abc import ABC, abstractmethod
+from types import TracebackType
+from typing import Generator, Optional, Type, Union
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+
+class BaseFOV(ABC):
+    # NOTE: not using the `data` method from ngff Position
+
+    @property
+    @abstractmethod
+    def channels(self) -> list[str]:
+        raise NotImplementedError
+
+    def channel_index(self, key: str) -> int:
+        """Return index of given channel."""
+        return self.channels.index(key)
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, int, int, int, int]:
+        # NOTE: suggestion, fix dimension to 5?
+        # We could me more restrictive than ome-zarr
```

Comment:
If we fix this to a 5-tuple, should `__getitem__` match this behavior? i.e. if an FOV has an underlying array of 3 dimensions, should the slicing be `[0,0,0]` or `[0,0,0,0,0]` to get a scalar?

Reply:
From the perspective of non-expert end users, I think it would be confusing if the pixel values in a 3D image had to be accessed using something like `[x, y, z, 0, 0]`. (It would be even worse if the placeholder channel and time dimensions came first, because then a naive indexing using `[x, y, z]` would not only not result in a scalar but would be entirely incorrect)

Reply:
Fair enough!

Reply:
There is a glitch in slicing with all 5 dimensions though. Translating the slicing makes it impossible to return a lazy Zarr array, removing many useful attributes (append, chunks, fancy slicing etc) that we currently inherit from `zarr.Array`. And at writing time, a NumPy array is directly written to Zarr, preserving shape, and it would be very confusing to have to slice them differently (or even having two different array types depending on how the file is opened).

Reply:
Uhm... an alternative is to implement a method on `BaseFOV` that translate the 5-dim indexing to the original array, and when requesting the whole array we can call `np.expand_dims`.

Reply:
> Uhm... an alternative is to implement a method on `BaseFOV` that translate the 5-dim indexing to the original array, and when requesting the whole array we can call `np.expand_dims`.

Yeah I think this is reasonable. It actually has to call `expand_dims` even when not requesting the whole array. So we might as well just delegate this to dask or tensorstore by constructing the reshaped always-lazy da/ts array with `dask()` or `tensorstore()` as we currently have with `ngff.ImageArray.numpy()`.

Reply:
> So we might as well just delegate this to dask or tensorstore by constructing the reshaped always-lazy da/ts array with `dask()` or `tensorstore()` as we currently have with `ngff.ImageArray.numpy()`.

This way we can also wrap dask in xarray and delegate the dimension naming API too.

Reply:
@JoOkuma and I had a chat and we will try out zarr's `view` for this.